Allow to use HTML5 audio tag with API Web Audio

marizuccara opened the pull request:
Introduced the WebAudioMedia backend which allows to load audio as HTML5 audio tag and use it with WebAudio backend. 

Setting the WebAudioMedia backend,there is a possibility to load audio of big dimensions, using the API Web Audio features. In this way, filters and other functionalities can be performed as with WebAudio backend without decoding internally audio data, that caused crash of the browser. 

### Breaking in the external API:

Added the WebAudioMedia backend. The API to use to load an audio or playback it, are the same as with the MediaElement backend. 

### Breaking changes in the internal API:

Added a file WebAudioMedia  that extends WebAudio. Instead of using the AudioBufferSourceNode to push the source to the AudioContext, it was used the MediaElementSourceNode which takes in input the audio tag. Since an HTML5 file is loaded, there were used APIs of MediaElement backend. 

### Todos/Notes:

It was created another backend to test this functionality and show you my proposal. In fact, I have duplicated the MediaElement code because the resource is handled as HTML5 audio tag. For other operations, there were used WebAudio APIs. 

### Related Issues and other PRs:

fixes #1763

katspaugh commented on src/webaudio-mediasource.js:
```diff
@@ -0,0 +1,459 @@
+        });
+    }
+    /** @private */
+    createScriptNode() {
```
Isn't this method exactly the same in the parent WebAudio class?

marizuccara replied:
Yes, i've added only createSource method which allow me to create MediaElementSourceNode (instead of AudioBufferSourceNode) and handle audio resource with API Web Audio. I've duplicated some code, only to show you my solution in a different file,  to give more clarity to my purpose.

aburai replied:
i worry about if we need a new backend to handle this, i need time to check it

katspaugh commented on src/webaudio-mediasource.js:
```diff
@@ -0,0 +1,459 @@
+     *
+     * @private
+     */
+    createVolumeNode() {
```
This one can also be inherited w/o duplication.

katspaugh commented on src/webaudio-mediasource.js:
```diff
@@ -0,0 +1,459 @@
+        this.gainNode.connect(this.ac.destination);
+    }
+    /** @private */
+    createAnalyserNode() {
```
Same here. Probably some others, too. I didn't check further.

marizuccara commented:
> Thanks for this PR!
> What was your incentive to implement this new backend?
> My guess would be some people might want to use a media element but control it via Web Audio?

If I load a big audio file with WebAudio backend, because it internally decodes the AudioBuffer, the browser crashes. Using HTML5 tag, the audio data are not decoded, so in case of a big audio file browser does not crash. Moreover, in this way I can use it with API Web Audio and add filters, like control left and right channels, pannerNode, equalizer, etc

katspaugh commented:
> Using HTML5 tag, the audio data are not decoded, so in case of a big audio file browser does not crash.

Even with HTML5, the audio data is still [decoded](https://github.com/katspaugh/wavesurfer.js/blob/master/src/wavesurfer.js#L1411-L1432) to draw the waveform.

marizuccara commented:
> Even with an HTML5 media element, the audio data is still [decoded](https://github.com/katspaugh/wavesurfer.js/blob/master/src/wavesurfer.js#L1411-L1432) to draw the waveform.

But giving peaks, the waveform is drawn with them

katspaugh commented:
Yes, correct.

marizuccara commented:
> Yes, correct.
With WebAudio backend, even if i give peaks to draw the waveform, it internally decode the audio to transform it to an AudioBuffer and put it to AudioBufferSourceNode. So the waveform is drawn without any problems, but in case of big audio files, the browser crashes because it decodes the data in memory. 
If we give an HTML5 tag with peaks, the waveform is drawn without decoding. Moreover, giving the source (HTML tag) to MediaElementSourceNode, it does not decode audio data in memory, so it can be played without any problems (like with MediaElement backend). But in this way we can add filters and have more controls with audio, because we can use API Web Audio

katspaugh commented:
Cool, gotcha.
I guess this new backend a good option then. More powerful than a standalone media element, but more efficient than the buffer source.
Let's clean up the PR and merge it. Any idea how to get that coverage back to 79%?

katspaugh commented on src/wavesurfer.js:
```diff
@@ -753,7 +759,6 @@ export default class WaveSurfer extends util.Observer {
         this.fireEvent('interaction', () => this.play(start, end));
         return this.backend.play(start, end);
     }
-
```
I suggest we restore this new line to avoid unnecessary changes.

katspaugh commented on src/wavesurfer.js:
```diff
@@ -1407,7 +1415,6 @@ export default class WaveSurfer extends util.Observer {
             });
         }
     }
-
```
Same here.

marizuccara commented:
> Cool, gotcha.
> I guess this new backend a good option then. More powerful than a standalone media element, but more efficient than the buffer source.
> Let's clean up the PR and merge it. Any idea how to get that coverage back to 79%?

Maybe, we can remove private to functions like createScriptNode, createVolumeNode and Analyser, so we can extend them normally

aburai commented on src/webaudio-mediasource.js:
```diff
@@ -0,0 +1,459 @@
+        /** @private */
+        this.params = params;
+        /** @private */
+        this.params = params;
```
duplicated code

marizuccara replied:
How does this work? Can I modify code deleting duplicated code or in this phase do you make optimizations?

katspaugh replied:
You can commit changes in your local repo and push to your GitHub fork's master.

katspaugh commented on src/mediaelement.js:
```diff
@@ -178,6 +184,15 @@ export default class MediaElement extends WebAudio {
+        this.sourceMediaElement = this.ac.createMediaElementSource(
+            mediaElement
+        );
+        this.sourceMediaElement.connect(this.analyser);
```
An alternative option would be to keep WebAudioMedia as a separate backend and just inherit from the MediaElement backend. With just these two methods (`init` and `createMediaElementSource`).

marizuccara replied:
Yes, it can be also an alternative if we want handle audio with MediaElement backend only as HTML5 tag.  It is a business logic choice, I can restore to WebAudioMedia backend and inherit from MediaElement!

katspaugh replied:
I think it would be cleaner as a separate class. 👍

marizuccara replied:
But putting createMediaElementSource to this new class, i have to override _load method because it uses createMediaElementSource method.

katspaugh replied:
In ES6, you can use [`super`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super) to refer to the parent class methods.
So your method would looks smth like:
```
_load() {
  super._load();
  this.createMediaElementSource(media);
}
```

marizuccara replied:
ah thanks!

thijstriemstra commented on src/webaudio.js:
```diff
@@ -135,9 +128,7 @@ export default class WebAudio extends util.Observer {
-        /** @private */
         this.scriptNode = null;
         /** @private */
         this.source = null;
```
@marizuccara can you replace `/** @private */` with a meaningful comment for each variable?

thijstriemstra replied:
can you also add `@since` in new classes and/or methods? https://jsdoc.app/tags-since.html

katspaugh commented:
We just need to add the new backend type to the docs now. The docs are on another branch, [`gh-pages`](https://github.com/katspaugh/wavesurfer.js/tree/gh-pages), so it should be another PR.

marizuccara commented:
> can you also look at simple unit tests (see existing MediaElement tests)?

Have I run your karma tests or write other tests?

marizuccara commented:
> > > can you also look at simple unit tests (see existing MediaElement tests)?
> > 
> > 
> > Have I run your karma tests or write other tests?
> 
> before pushing pull requests one should always run the tests with `npm run test` or `yarn test`
> but in this case we add something new that should have test coverage
> an example is found under `spec/mediaelement.spec.js`
> if you need any help, let me know, when i have time to test the new backend, i can have a look at the tests and/or append some tests that make sense

Thanks. I've created  `spec/webaudio-media.spec.js` which has the same tests like MediaElement backend. These are the results: 

 
![image](https://user-images.githubusercontent.com/44341274/66656311-0656f380-ec3e-11e9-8c7b-90c6f02336b6.png)

thijstriemstra commented:
thanks @marizuccara can you commit those as well so the CI service can run the tests as well?

thijstriemstra commented on src/wavesurfer.js:
```diff
@@ -1322,6 +1329,8 @@ export default class WaveSurfer extends util.Observer {
                 return this.loadBuffer(url, peaks, duration);
             case 'MediaElement':
                 return this.loadMediaElement(url, peaks, preload, duration);
```
this line can be removed if it's the same call for both cases.

thijstriemstra commented on src/webaudio-media.js:
```diff
@@ -0,0 +1,56 @@
+import MediaElement from './mediaelement';
+
+/**
+ * WebAudioMedia backend
```
can you add a description for the backend here (possibly copied from the discussion on this PR)?

thijstriemstra commented on src/webaudio-media.js:
```diff
@@ -0,0 +1,56 @@
+
+/**
+ * WebAudioMedia backend
+ * @since Version 3.1.0
```
new version would be 3.2.0

thijstriemstra commented on src/webaudio-media.js:
```diff
@@ -0,0 +1,60 @@
+    /**
+     * Create MediaElementSource node
+     *
+     * @since Version 3.1.0
```
incorrect version nr

marizuccara replied:
Ah, sorry, I forgot to edit this version number!
I've documented also `mediaType` and `mediaControls` properties.

thijstriemstra replied:
only the version number is ok, don't include `Version`

thijstriemstra commented:
>      Thanks. I've created spec/webaudio-media.spec.js which has the same tests like MediaElement backend.

Having a base test class that can be used for both mediaElement and this new backend would be useful. Thoughts?

marizuccara commented:
> > ```
> >  Thanks. I've created spec/webaudio-media.spec.js which has the same tests like MediaElement backend.
> > ```
> 
> Having a base test class that can be used for both mediaElement and this new backend would be useful. Thoughts?
> 
> Also: will `wavesurfer.loadBlob` work with this backend?

No, `wavesurfer.loadBlob` does not work with this backend because it does not decode audio data, and with `wavesurfer.loadBlob` audio data are decoded. So this method works only with WebAudio backend.

marizuccara commented:
> > > ```
> > >  Thanks. I've created spec/webaudio-media.spec.js which has the same tests like MediaElement backend.
> > > ```
> > 
> > 
> > Having a base test class that can be used for both mediaElement and this new backend would be useful. Thoughts?
> > Also: will `wavesurfer.loadBlob` work with this backend?
> 
> No, `wavesurfer.loadBlob` does not work with this backend because it does not decode audio data, and with `wavesurfer.loadBlob` audio data are decoded. So this method works only with WebAudio backend.

Maybe is it possible to implement a shared file which has the tests for both backends and include this file to `mediaelement.spec` and `webaudio-media.spec`?

marizuccara commented:
> can you also add an example that uses the backend? Also take a look at adding it to the changelog

Added examples and updated CHANGES.md. 
Tell me what you think about the solution of shared tests.

thijstriemstra commented:
I'm not sure about the name `WebAudioMedia`, it's a term we came up and I find it not clear. What about `MediaElementSource` or something related to the method we're using?

marizuccara commented:
> I'm not sure about the name `WebAudioMedia`, it's a term we came up and I find it not clear. What about `MediaElementSource` or something related to the method we're using?
> 
> > updating changelog caused conflicts
> 
> Fixed this for you.
> 
> > Tell me also what do you think about the solution of shared tests.
> 
> Looks good.

Thanks! I called it in this way for using mediaelement with Api Web Audio. With MediaElementSource is it clear that i can use it with API Web Audio, since the creation of the source is an internal stuff? But anyway, surely you know better what is the best way to call this backend, so if you think that is clearer calling it MediaElementSource, i can change without any problems the name!

thijstriemstra commented:
> With MediaElementSource is it clear that i can use it with API Web Audio,

No, not really. Maybe `MediaElementWebAudio` then? It's long but it's clear (to me).

marizuccara commented:
> > With MediaElementSource is it clear that i can use it with API Web Audio,
> 
> No, not really. Maybe `MediaElementWebAudio` then? It's long but it's clear (to me).

Yes, I also think that this new name is more explicit and sounds better! So I'm changing it, okay?

thijstriemstra commented:
> So I'm changing it, okay?

Yep!

thijstriemstra commented on src/wavesurfer.js:
```diff
@@ -30,12 +31,19 @@ import PeakCache from './peakcache';
  * @property {string} backgroundColor=null Change background color of the
  * waveform container.
- * @property {number} barHeight=1 The height of the wave bars.
+ * @property {number} barHeight=1 The height of the wave bars.F
```
typo

thijstriemstra commented on src/wavesurfer.js:
```diff
@@ -30,12 +31,19 @@ import PeakCache from './peakcache';
+ * MediaElementWebAudio allows to use API Web Audio also with big audio files, loading audio like with
+ * MediaElement backend (HTML5 audio tag). You have to use the same methods of MediaElement backend for loading and
+ * playback, giving also peaks, so the audio data are not decoded. In this way you can use WebAudio features, like filters,
+ * also with audio with big duration.For example:
```
Missing space before `For`

thijstriemstra commented on src/wavesurfer.js:
```diff
@@ -67,10 +75,10 @@ import PeakCache from './peakcache';
- * @property {string} mediaType='audio' (Use with backend `MediaElement`)
- * `'audio'|'video'`
+ * @property {string} mediaType='audio' (Use with backend `MediaElement` or `MediaElementWebAudio`)
+ * `'audio'|'video'` ('video' only for MediaElement)
```
missing backticks

thijstriemstra commented on src/webaudio.js:
```diff
@@ -11,9 +11,9 @@ const FINISHED = 'finished';
+    /** scriptBufferSize: size of the processing buffer */
     static scriptBufferSize = 256;
-    /** @private */
+    /** audioContext: allows to process audio with API Web Audio */
```
"WebAudio API"

thijstriemstra commented on spec/mediaelement-mediaelement_webaudio-shared.js:
```diff
@@ -0,0 +1,275 @@
+/* eslint-env jasmine */
```
why not use a filename like `media-element-shared.js`?

marizuccara commented:
The new tests that you write for MediaElement backend (error tests), maybe are good also for the new backend since we can pass to load method the HTML5 tag?

marizuccara commented:
> > Added examples and updated CHANGES.md (updating changelog caused conflicts).
> 
> I don't see the example?

marizuccara commented on CHANGES.md:
```diff
@@ -5,6 +5,16 @@ wavesurfer.js changelog
+    ` wavesurfer.load(url, peaks, preload, duration);
+      wavesurfer.play();
+      wavesurfer.setFilter(customFilter);
+    `
```
This example is not good? The same in changelog

thijstriemstra replied:
I was thinking of an example in the `example` directory that uses the new backend (and adds some filters for example).

marizuccara replied:
I've pushed the example.
- The filters provided are good? In the explanation i've commented only one because took me a lot of space.
- I've uploaded a file audio taken from `split-channels` example. I've initially wanted to load a 3 hours audio (because with an audio of this dimensions, with `WebAudio` backend the browser crashes, but with `MediaElementWebAudio` not, but GitHub tell me that the file is too large).
- I've used `audiowaveform` with `split-channels` command, but the data array that returned is only of one dimension, so i drawn the entire waveform, without splitting the channels (if you know how i can solve the problem, tell me)

thijstriemstra commented on example/mediaelement-webaudio/app.js:
```diff
@@ -0,0 +1,91 @@
+            return response.json();
+        })
+        .then(peaks => {
+            let max = peaks.data.reduce((max, el) => (el > max ? el : max));
```
can you add a comment for the user what's happening here?

thijstriemstra commented on example/mediaelement-webaudio/app.js:
```diff
@@ -0,0 +1,91 @@
+        })
+        .then(normalizedPeaks => {
+            // Load audio from URL
+            wavesurfer.load('stereo.mp3', normalizedPeaks, 51);
```
you should load the file from `../media/`. If it's a copy, adjust both examples to use the same file.

thijstriemstra commented on example/mediaelement-webaudio/peaks.json:
```diff
@@ -0,0 +1 @@
+{"version":2,"channels":2,"sample_rate":44100,"samples_per_pixel":256,"bits":8,"length":8861,"data":[-5,7,-2,2,-5,7,-2,2,-4,6,-1,0,-4,6,-1,1,-4,6,-2,1,-4,5,-2,1,-3,4,-2,1,-2,4,-1,1,-1,3,-1,0,0,2,0,0,0,1,-1,1,0,2,-1,1,-1,3,-1,1,-2,4,-1,1,-3,4,-1,1,-3,5,-1,0,-3,5,-1,1,-3,5,-1,1,-2,5,-1,1,-1,4,-1,1,0,4,-1,0,0,3,-1,0,0,3,0,0,-4,8,-4,3,-5,9,-4,3,-5,8,-4,3,-3,7,-3,2,-4,7,-3,2,-4,7,-3,2,-3,7,-3,2,-2,6,- ...
```
i'd call it 'filename-peaks' so `stereo-peaks.json`.

kavishatalsania commented:
Hi @marizuccara / @thijstriemstra / @katspaugh ,
Thanks for this great feature.

I used 'MediaElementWebAudio' backend. It does not load whole audio and I am also able to use filters to the mute channel.

One issue I found is when I load the audio file from s3 it is giving error **'MediaElementAudioSource outputs zeroes due to CORS access restrictions**'

I am able to resolve this issue by adding line media.crossOrigin = "anonymous" in the load method of wavesurfer at line 1955.

```
key: "load",
value: function load(url, container, peaks, preload) {
      var media = document.createElement(this.mediaType);
      media.controls = this.params.mediaControls;
      media.autoplay = this.params.autoplay || false;
      media.preload = preload == null ? 'auto' : preload;
      media.src = url;
      media.style.width = '100%';
      media.crossOrigin = "anonymous";
      var prevMedia = container.querySelector(this.mediaType);
      if (prevMedia) {
        container.removeChild(prevMedia);
      }
      container.appendChild(media);
      this._load(media, peaks);
}
```
Is it right resolution for this problem?

marizuccara commented:
You should be able also to create yourself an audio tag.
For example, `<audio src="http://domain/audio" crossOrigin="anonymous">`,
or by JavaScript, 
` let audio = document.createElement('audio');`
`audio.src = "http://domain/audio";`
`audio.crossOrigin = "anonymous";`
Then, pass the audio tag to `load` method:
`wavesurfer.load(audio, peaks)`.

kavishatalsania commented:
Hi @marizuccara 
Thanks for your suggestion. It's working.

I got impression from the documentation and code that I could only pass the audio URL in the load method not audio object.

marizuccara commented on CHANGES.md:
```diff
@@ -18,6 +17,7 @@ wavesurfer.js changelog
       wavesurfer.play();
```
Maybe in this way is clearer?
`wavesurfer.load(url | HTMLMediaElement, peaks, preload, duration);`

kavishatalsania replied:
True

thijstriemstra commented:
> audio.crossOrigin = "anonymous";

can we add this line to the example?

thijstriemstra commented:
merged. thanks @marizuccara!

katspaugh commented:
Thanks a lot, @marizuccara! A great contribution!

marizuccara commented:
Thanks so much to you, @katspaugh and @thijstriemstra, for this fantastic library!

Leaybc commented:
Looking forward to releasing new version as soon as possible

kavishatalsania commented:
Hi @katspaugh / @thijstriemstra 

When are you planning to release new version with these changes?
